src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentFactoryBase.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage
+{
+    internal abstract class VirtualDocumentFactoryBase : VirtualDocumentFactory
+    {
+        private readonly IContentTypeRegistryService _contentTypeRegistry;
+        private readonly ITextBufferFactoryService _textBufferFactory;
+        private readonly ITextDocumentFactoryService _textDocumentFactory;
+        private readonly FileUriProvider _fileUriProvider;
+        private IContentType _languageLSPContentType;
+
+        public VirtualDocumentFactoryBase(
```

Comment:
Given this class is now a shared for many factories there'd probably be value in unit testing it to ensure things behave as expected and nothing breaks any of the base expectations for any of us.

Reply:
Are there existing tests we want to move, or do we want to add new ones? 

---
In reply to: [451976681](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r451976681) [](ancestors = 451976681)

Reply:
New ones. In the past the C#/HTML virtual document factories were iterating so frequently we didn't add tests for them but things have settled down now

Reply:
Sounds good, will do.

---
In reply to: [452411026](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r452411026) [](ancestors = 452411026)

Reply:
Actually there were some language-specific ones already. I ported them to be generic, but left language-specific ones in place since they cover a bit more, including language content types.

---
In reply to: [452474190](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r452474190) [](ancestors = 452474190,452411026)